processor/spanmetricsprocessor/README.md

```diff
@@ -0,0 +1,106 @@
+promexample_latency_bucket{http_method="GET",http_status_code="200",label1="value1",operation="/Address",service_name="shippingservice",span_kind="SPAN_KIND_SERVER",status_code="STATUS_CODE_UNSET",le="10"} 1195
+promexample_latency_bucket{http_method="GET",http_status_code="200",label1="value1",operation="/Address",service_name="shippingservice",span_kind="SPAN_KIND_SERVER",status_code="STATUS_CODE_UNSET",le="100"} 10180
+promexample_latency_bucket{http_method="GET",http_status_code="200",label1="value1",operation="/Address",service_name="shippingservice",span_kind="SPAN_KIND_SERVER",status_code="STATUS_CODE_UNSET",le="250"} 10180
+...
+```
+
+Each metric will have _at least_ the following dimensions because they are common across all spans:
+- Service name
+- Operation
+- Span kind
+- Status code
+
+This processor lets traces to continue through the pipeline unmodified.
+
+The following settings are required:
+
+- `metrics_exporter`: the name of the exporter that this processor will write metrics to. This exporter **must** be present in a pipeline.
+
+The following settings can be optionally configured:
+
+- `latency_histogram_buckets`: the list of durations defining the latency histogram buckets.
+  - Default: `[2ms, 4ms, 6ms, 8ms, 10ms, 50ms, 100ms, 200ms, 400ms, 800ms, 1s, 1400ms, 2s, 5s, 10s, 15s]`
+- `dimensions`: the list of dimensions to add together with the default dimensions defined above. Each additional dimension is defined with a `name` which is looked up in the span's collection of attributes. If the `name`d attribute is missing in the span, the optional provided `default` is used. If no `default` is provided, this dimension will be **omitted** from the metric.
+
+Example:
+
+```yaml
+receivers:
+  jaeger:
+    protocols:
+      thrift_http:
+        endpoint: "0.0.0.0:14278"
+
+  # Dummy receiver that's never used, because a pipeline is required to have one.
+  otlp/spanmetrics:
+    protocols:
+      grpc:
+        endpoint: "localhost:12345"
+
+  otlp:
+    protocols:
+      grpc:
+        endpoint: "localhost:55677"
+
+processors:
+  spanmetrics:
+    metrics_exporter: otlp/spanmetrics
+    latency_histogram_buckets: [2ms, 6ms, 10ms, 100ms, 250ms]
+    dimensions:
+      - name: http.method
+        default: GET
+      - name: http.status_code
+
+exporters:
+  jaeger:
+    endpoint: localhost:14250
+
+  otlp/spanmetrics:
+    endpoint: "localhost:55677"
+    insecure: true
+
+  prometheus:
+    endpoint: "0.0.0.0:8889"
+    namespace: promexample
+
+pipelines:
+  traces:
+    receivers: [jaeger]
+    processors: [spanmetrics, batch, queued_retry]
+    exporters: [jaeger]
+
+  # The exporter name must match the metrics_exporter name.
+  # The receiver is just a dummy and never used; added to pass validation requiring at least one receiver in a pipeline.
+  metrics/spanmetrics:
```

Comment:
> Processors allow the flexibility of writing to a metrics exporter without the overhead of an additional pipeline, if no processing of metrics is required. An example is exporting span metrics directly to prometheus.

Just wondering, how much overhead are we expecting for another pipeline? Isn't it just running `consumeTraces` twice instead of once? Or, I've never tried, but is it not allowed to have the same receiver in multiple pipelines? That seems like it would make the config really understandable.

```
pipelines:
  traces:
    receivers: [jaeger]
    processors: [batch, queued_retry]
    exporters: [jaeger]

  tracemetrics:
    receivers: [jaeger]
    processors: [spanmetrics]
    exporters: [prometheus]
```

This is similar to the translator approach, but seems more intuitive - just let receivers take part in multiple pipelines without any other type of component, the collector should be able to manage wiring a receiver to multiple sinks without a big problem and seems like something that should be supported anyways since users will probably try this sort of config (I haven't tried it yet, does it really not work?).

I think this is another argument to start with this current implementation though :) But just wondering if this idea has come up.

Reply:
Thanks for your thoughts, @anuraaga.

> Or, I've never tried, but is it not allowed to have the same receiver in multiple pipelines?

Yup :) 

> how much overhead are we expecting for another pipeline? Isn't it just running consumeTraces twice instead of once?

There's some potential performance overhead if an earlier pipeline's first processor is blocked or slow. From the [OTEL Design doc](https://github.com/open-telemetry/opentelemetry-collector/blob/master/docs/design.md):

> The data propagation from receiver to FanOutConnector and then to processors is via synchronous function call. This means that if one processor blocks the call the other pipelines that are attached to this receiver will be blocked from receiving the same data and the receiver itself will stop processing and forwarding newly received data.

I like your proposed config, it does look very intuitive.

However, pipelines don't support mixed telemetry types right now; if this were possible, I think your suggestion makes most sense. Mixed telemetry type pipelines is something I'd like to tackle once the `spanmetricsprocessor` is done, because it sounds like a fairly significant change.

This is the error message I get if I use a `jaeger` receiver with `prometheus` exporter:
```
application run finished with error: cannot setup pipelines: cannot build builtExporters: pipeline "traces" of data type "traces" has an exporter "prometheus", which does not support that data type
```